### ISSUE TITLE
About FVD computing

### ISSUE THREAD

**Issue by johannwyh** [sihyun-yu/digan#3]
Hello! I have two questions about FVD computing.

* `frechet_video_distance.py` forces the generated sequence to be of length 16
```python
fake = torch.cat([rearrange(
    G(z, c, timesteps=16, noise_mode='const')[0].clamp(-1, 1).cpu(),
    '(b t) c h w -> b t h w c', t=16) for z, c in zip(grid_z, grid_c)])
```
If I want to train DIGAN with clips length of 32 or 128. What should I do for FVD computing? 
* How long does it take to compute FVD each time? Following your training setting, this process takes 30 minutes each time and is calculated every 400k image, which means that the time spent on FVD calculation will be 0.5hrs * 25000 / 400 = 32hrs. This is far more time-consuming than StyleGAN2 FID computing. I want to make sure is this the case when you train your own model?

Thank you very much!

**Comment by sihyun-yu** [comment]
You can set longer length for FVD calculation (while training) if the model with longer clips, but note that it will consume more memory and time. 

For FVD calculation, it seems quite strange that the process takes about 30 minutes each time. In our case, it takes ~6 minutes for each calculation while training.

**Comment by johannwyh** [comment]
Thanks for your feedback! I am a little confusing on the `compute_fvd` function

``` python
def compute_fvd(opts, num_videos=512, num_iter=5):
    start_time = time.time()
    # load vid classifier
    i3d = _load_i3d_pretrained(opts.device)

    # load dataloader
    dataset = dnnlib.util.construct_class_by_name(return_vid=True,
                                                  **opts.dataset_kwargs)  # return_vid True for FVD measure

    # Copy generator
    G = copy.deepcopy(opts.G).eval().requires_grad_(False).to(opts.device)

    # compute fvd
    fvd = [_eval_fvd(opts, i3d, G, dataset, opts.device, num_videos=num_videos) for _ in range(num_iter)]

    if opts.rank != 0:
        return float('nan')

    return sum(fvd)/float(num_iter)
```

For all subprocess with non-zero `rank`, the function does a lot of computation and finally returns a `nan`. I wonder whether the `nan` can be returned at the beginning of the function? 

BTW, is this consistent with your implementation and can be calculated within 6 minutes?

Thanks a lot!

**Comment by johannwyh** [comment]
I find my own dataset is all of 1024 resolution. Will it be the cause of the low speed of FVD computation? 

I am resizing the dataset to 128x128, perhaps the data-loading process in _eval_fvd may get faster.

**Comment by sihyun-yu** [comment]
I think it depends on your CPUs. Can you try turning on `pin_memory=True` or `num_workers>0`?

**Comment by johannwyh** [comment]
Hi, I made some experiments and I want to report some results for your information.

- I switched from `PIL.Image.read` to `pyspng.load`, the per video loading time reduces from 2.6s to 0.5s. For pyspng to work, [this line](https://github.com/sihyun-yu/digan/blob/8368d5bfd73db38a9593ce10bb246cd7c37ddf9f/src/training/dataset.py#L650) of `dataset.py` should be modified to
```python
img = rearrange(img, 'h w c -> c h w').astype(np.float32)
```
Otherwise, the read image is of *int* type and will cause an error on `F.interpolate`

- I set `pin_memory=False` for the fvd dataloader and the per video loading time reduces from 0.5s to 0.3s
- After gathering of embeddings, I only compute fvd when `opts.rank==0`

Now the fvd computation speed achieves your standard.

Finally, thank you for your instant feedbacks for all these time!!!!!
